src/main/java/com/powsybl/openloadflow/ac/nr/PerEquationTypeStoppingCriteria.java

```diff
@@ -0,0 +1,113 @@
+package com.powsybl.openloadflow.ac.nr;
+
+import com.powsybl.openloadflow.ac.equations.AcEquationType;
+import com.powsybl.openloadflow.ac.equations.AcVariableType;
+import com.powsybl.openloadflow.equations.EquationSystem;
+import com.powsybl.openloadflow.equations.Vectors;
+import com.powsybl.openloadflow.util.PerUnit;
+
+/**
+ * @author Alexandre Le Jean <alexandre.le-jean at artelys.com>
+ */
+public class PerEquationTypeStoppingCriteria implements NewtonRaphsonStoppingCriteria {
+
+    // FIXME Any idea for the name?
+    private final double defaultValueAlpha = Math.pow(10, -5);
+
+    // FIXME Any idea for the name?
+    private final double defaultValueRho = Math.pow(10, -5);
+
+    // FIXME Any idea for the name?
+    private final double defaultValueB = Math.pow(10, -5);
+
+    // FIXME Any idea for the name?
+    private final double defaultValuePhi = Math.pow(10, -5);
+
+    private final double maxActivePowerMismatch;
+
+    private final double maxReactivePowerMismatch;
+
+    private final double maxVoltageMismatch;
+
+    public PerEquationTypeStoppingCriteria(double maxActivePowerMismatch,
+                                           double maxReactivePowerMismatch, double maxVoltageMismatch) {
+        this.maxActivePowerMismatch = maxActivePowerMismatch;
+        this.maxReactivePowerMismatch = maxReactivePowerMismatch;
+        this.maxVoltageMismatch = maxVoltageMismatch;
+    }
+
+    private double computeNorm(double[] fx) {
+        return Vectors.norm2(fx);
+    }
+
+    private boolean computeStop(double[] fx, EquationSystem<AcVariableType, AcEquationType> equationSystem) {
+        for (var eq : equationSystem.getIndex().getSortedEquationsToSolve()) {
+            var type = eq.getType();
+            var idx = eq.getColumn();
+            switch (type) {
+                case BRANCH_TARGET_P:
+                case BUS_TARGET_P:
+                case DUMMY_TARGET_P:
+                    if (Math.abs(fx[idx]) * PerUnit.SB >= maxActivePowerMismatch) {
+                        return false;
+                    }
+                    break;
+                case BRANCH_TARGET_Q:
+                case BUS_TARGET_Q:
+                case DISTR_Q:
+                case DUMMY_TARGET_Q:
+                    if (Math.abs(fx[idx]) * PerUnit.SB >= maxReactivePowerMismatch) {
+                        return false;
+                    }
+                    break;
+                case BUS_TARGET_V:
+                case BUS_TARGET_V_WITH_SLOPE:
+                case ZERO_V:
+                    if (Math.abs(fx[idx]) >= maxVoltageMismatch) {
+                    // FIXME, condition must be : Math.abs(fx[idx]) * PerUnit.zb(idx_nominal_voltage) >= maxVoltageMismatch
+                    //  We need the value NominalVoltage associated to the equation
+                    //  either get the value or the associated BUS to fetch the information
+                        return false;
+                    }
+                    break;
```

Comment:
I might disagree with the FIXME and leave maxVoltageMismatch in p.u.

Also Entsoe uses this parameter in pu, see [here](https://eepublicdownloads.azureedge.net/clean-documents/digital/QualityOfCGMESdatasetsAndCalculations_v3_3.pdf) line 5567

Reply:
Ok then. You can leave voltage mismatch in pu and put default threshold to 10^-4 as in this document.